Add preliminary networking spec.

krystynneisess opened the pull request:
As requested by @yizow

krystynneisess committed:
[NETWORKING] Add preliminary networking spec.

akumaraguru commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,79 @@
+
+The FC network will have up to eight clients, as follows:
+
+- One FC client.  This is the FC computer running the match.  A wifi-enabled PiE computer+monitor. 
```
There's no reason for the FC client or driver stations to be on wireless connections. Probably want to remove "wifi enabled" from the spec so we don't give the wrong idea.

akumaraguru commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,79 @@
+- Four driver station clients.  This acts as the student-robot interface (via Dawn and Runtime), and as the robot-field control interface (see "Establishing Communication Links").  A wifi-enabled laptop or  computer+monitor.
+- Up to four robot clients.  A wifi-enabled Beagle Bone Black (via wifi dongle).  One per student robot.
+
+### Assigning IP Addresses
```
How is the network supposed to handle IP address conflicts? Poor management of Beaglebones may cause this.

akumaraguru commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,79 @@
+Within each match, the FC network will create/maintain eight TCP connections:
+
+- A connection between the FC client and each Driver Station client.  These can persist across matches, and so need to be opened only once per network initialization.
+- A connection beween a Driver Station client and its corresponding Robot client. These must be opened/closed at the start/end of each match (i.e. once the Robot client joins/leaves the FC network and is paired/unpaired with a Driver Station client).
```
Might want to specify who (FC client, Dawn, ???) holds responsibility for when a TCP connection is broken.

nikitakit replied:
So I believe Dawn<->Runtime connection is primarily UDP. (though if you need the TCP endpoint as well on the field, the spec should accommodate both)

And yes, you need to spell out the behavior in case:
* The robot restarts, thus resetting the connection
* The driver station is unplugged from the ethernet cable, thus resetting the connection (yes, this has actually happened on the competition field)
* The connection enters an invalid state, thus necessitating a reset.

Alramech replied:
There also should be a specification of how FC knows which driver station gets paired with which robot. Possibilities include reading from the spreadsheet, manual input, or some other method of communication

akumaraguru reviewed (changes requested):
This is good for giving general software folk an overview of how the network is designed. Idk if it's meant for anything else, but if so, it would be good to have more deals on implementation.

krystynneisess committed:
[NETWORKING] Make minor spec revisions.

Correct client description; address IP collisions and TCP connection drops.

krystynneisess commented:
Comments addressed!

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+### Clients
+
+The FC network will have up to eight clients, as follows:
```
Nine total

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+Consider the following example static IP scheme ("P" = subnet prefix):
+
+- FCl:  P.255
```
Not a valid IP. This is the reserved broadcast IP

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+### Assigning IP Addresses
+
+Every client will have a statically assigned IP address.  This is feasible due to the small number of total clients (1 per robot ( less than 30 in total), 4 driver stations, and one field control center).
```
Add in specifics of IP. You can do this after taking 168.

Subnet, etc. Everything needed to configure the meraki router.

Alramech replied:
Since the robot connects to the student network and the field network over the same interface, specify whether they use the same IP for both networks or have independent ones

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+### Configuration
+
+A robot can connect to at most one network at a time, which must be either the Student network or the FC network.  The Student network is to be listed as higher priority than the FC Network.  This can be set each robot's network config file.
```
This can be set *in* each

nikitakit replied:
FYI if there is any need to update the network config file as part of runtime updates, I suggest using the lookup table trick I proposed in an earlier meeting.

* All robots receive a config file mapping from a unique ID (e.g. mac address), to any robot-specific settings like the IP
* All robots receive a script that reads the config file and the unique ID to generate the config file and put it in the right place

This means that if we need to update the network config files for any reason, we can easily deploy an update to all robots without breaking their functionality.

(On that note, if we go with this scheme -- this spec should mention the IPs that are reserved for staff-internal beaglebones)

nikitakit replied:
An example of one reason we'd need to deploy a custom config: say we need to update the settings for field control connections due to last-minute discovering issues with our field router config.

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+1. When not in a match, students will always be able to connect with their robots over the Student network even if the robot is within range of the FC network.
+2. When in a match, robots will only have the option of connecting to the FC network (due to router confiscation). 
+
+Note that this may result in robots not in a match to still be connected to the FC network.  However, this is totally fine since the FC client will not be listening to robots not currently tied to a driver station (see "Sending Messages").
```
Add how students will fix. 
namely, restarting robot.

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+The FC client sends messages to initiate robot-to-driver-station pairing, and to indicate what mode (autonomous, tele-op, etc) the robot is in.   Driver station clients listen accordingly, and pass the message to robots.
+
+Robot clients send only a heartbeat, which contains a status value that the FC client can use to verify the robot is both connected an in the expected mode.
```
Driver station sends heartbeat. 

Robot heartbeats are a runtime/dawn protocol, and outside the scope of this document.

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+### From FC
+
+**Robot-to-Driver-Station Pairing.**  Prior to this message, a robot, while connected to the network, is not yet associated with a driver station.  Containing the static IP of the Robot client with which to pair, the FC client sends this to the Driver Station client, which then opens a connection and begins communication with said robot.
```
We probably want to send just the robot number. 
The IP can be determined from the number.
In the future, we can use a lookup table if we don't want to use static IP.

nikitakit replied:
So the static IP can be deterministically generated from the match schedule. But should FC also do some checking first, e.g. pinging the robot?

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+**Robot-to-Driver-Station Pairing.**  Prior to this message, a robot, while connected to the network, is not yet associated with a driver station.  Containing the static IP of the Robot client with which to pair, the FC client sends this to the Driver Station client, which then opens a connection and begins communication with said robot.
+
+If more than one robot responds to the pairing message (i.e. IP address collision), or if no robot responds, FC is notified by the Driver Station.  The match will not proceed until this is externally resolved (e.g. reprogramming the suspected robot(s)'s Beaglebones).
```
Respond with what message? Different for each? Same for both error cases? 

Alternative: no special message gets sent, but FC station sees from heartbeat that robot remains disconnected.

Alramech replied:
How will detecting more than one robot be detected? What if 2 robots respond, but 1 is much later than the other?

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+If more than one robot responds to the pairing message (i.e. IP address collision), or if no robot responds, FC is notified by the Driver Station.  The match will not proceed until this is externally resolved (e.g. reprogramming the suspected robot(s)'s Beaglebones).
+
+**Mode Changing.**  This message contains an integer indicating whether the robot should be in Autonomous (1), Tele-Op (2), Disabled (3), or E-Stopped (4) mode.  Sent from FC to Driver Station, then Driver Station to Robot.
```
- Recommend zero indexing
- Recommend setting e-stop as 0

yizow commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+### To FC
+
+**Heartbeat.** Upon pairing, the robot begins sending a "heartbeat"--a message containing a single integer between 0 and 4 (inclusive) indicating its mode. If no mode change message has been received yet, it replies Connected (0) by default. Sent from Robot to Driver Station, then Driver Station to FC.
```
outside scope.

nikitakit commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+The FC network will have up to eight clients, as follows:
+
+- One FC client.  This is the FC computer running the match.  A PiE computer+monitor. 
+- Four driver station clients.  This acts as the student-robot interface (via Dawn and Runtime), and as the robot-field control interface (see "Establishing Communication Links").  A wifi-enabled laptop or  computer+monitor.
```
I don't think you should mention runtime in this bullet point.

nikitakit commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+- One FC client.  This is the FC computer running the match.  A PiE computer+monitor. 
+- Four driver station clients.  This acts as the student-robot interface (via Dawn and Runtime), and as the robot-field control interface (see "Establishing Communication Links").  A wifi-enabled laptop or  computer+monitor.
+- Up to four robot clients.  A wifi-enabled Beagle Bone Black (via wifi dongle).  One per student robot.
```
If you want to do video streaming, this may add an extra computer on the network. This adds some interesting constraints:
* Video streaming is hardware-demanding (running other things on same computer may not be a good idea)
* Video streaming needs a human to operate occasionally, so it's nice to be able to have one operator manage the FC base station (i.e. input match numbers), and a second person to oversee streaming.
* Video streaming needs access to the internet! (should be obvious why)
* Video streaming needs access to the field control internal network, so that it can get the current match timer in order to display it up on the screen.

The need for both Internet and internal networks complicates the setup somewhat.

nikitakit replied:
In addition to video streaming, the match schedule used to be pulled from an online source (e.g. a Google Spreadsheet), such that both event planners and field control operate off the same source of ground truth. I think one year we passed around schedules on paper slips and USB drives... that was a mess.

This also adds the need for a computer that has access to both field control and the internet.

nikitakit reviewed (dismissed):
Read my comments first.

In addition, the following communication modes we previously had in field control are absent from this spec:
* Sending the current match time from FC to Dawn (so students know how much time is left)
* Sending the current match time and teams involved such that they can be displayed in the video on the overhead projector (which may also be streamed)
* Note that e-stop should be triggerable from anywhere: from the robot (e.g. via a determination of unsafe battery buzzer voltage), from Dawn (via the student hitting e-stop), or from the field control operator (i.e. PiE field staff decide the robot is operating in an unsafe manner). Technically the from-robot functionality is not a networking thing because the robot can always decide for itself to permanently stay in e-stop mode.

Alramech commented on networking/Field_Control_Networking_Spec.md:
```diff
@@ -0,0 +1,83 @@
+
+## 5. Sending Messages
+
+The FC client sends messages to initiate robot-to-driver-station pairing, and to indicate what mode (autonomous, tele-op, etc) the robot is in.   Driver station clients listen accordingly, and pass the message to robots.
```
robot should initiate pairing via message to FC as it is generally unclear when it is ready to communicate unless inspecting every packet

krystynneisess committed:
[NETWORKING] Address comments parade.

Includes updates/corrections/additions regarding:

- Network IP configuration
- Network config scripts
- Driver Station heartbeats
- Video streaming accomodations
- Robot-Driver-Station pairing

As well as minor changes throughout.
Does not include proofreading for capitalization consistency and such.

nikitakit commented:
Umm, what's the state of this? It's been a while since the last comment or spec change.
